### ISSUE TITLE
PB626: Dictionary word not visibly highlighted after long tap and HW buttons not functional for "switching" between dictionaries

### ISSUE THREAD

**Issue by sonix-github** [koreader/koreader#7126]
* KOReader version: v2020.12-91-g1c9c40f_2021-01-07
* Device: PB626 sw:U626.5.20.1157

#### Issue
A)
Single word not visibly highlighted after long tap. It is highlighted in background right after dictionary window is showing the translation of selected word. 

B)
HW device buttons are not functional for "switching" between dictionaries. I have installed 3 dictionaries. If the dictionary lookup result is shown, it shows first translation. To the second/third dictionary I can no more use HW buttons, Only SW ones "<<" ">>" are functional. HW buttons are functional only when there is more information from dictionary lookup and the slider is shown. In this case I can move across text up/down also HW buttons.

#### Steps to reproduce
1. Open some e-book. I have pdf document.
2. Make long tap on some word - no word highlight is visible
3. First dictionary with lookup result for selected word is shown - in the background is now visible the highlighted word
4. Press up/down HW device buttons - not possible to move to next/previous dictionary

Before I used koreader nightly v2020.12-30-gb78d9c1_2020-12-23 and this issues were not there present.

Could you please check/fix it?
Thanks

**Comment by NiLuJe** [comment]
B/ Assuming you meant page turn buttons, those are eaten by the actual content (e.g., they should allow you to scroll inside the *current* definition).

I can confirm A/ on Kobo, although I rarely read PDFs, so I can't say if this is new or not. Works fine in CRe, though.

**Comment by Frenzie** [comment]
> I can confirm A/ on Kobo, although I rarely read PDFs, so I can't say if this is new or not. Works fine in CRe, though.

That's new. Pinging @poire-z

**Comment by poire-z** [comment]
Looking at A.

About B / 4:
> Press up/down HW device buttons - not possible to move to next/previous dictionary
> Before I used koreader nightly v2020.12-30-gb78d9c1_2020-12-23 and this issues were not there present

Are you sure about that ? Because I don't see they were ever made to switch to prev/next dict (but they were made to scroll up/down inside a defintion).
I guess they could be made to behave like that - but I want a confirmation it's not a regression :)

**Comment by Frenzie** [comment]
I think it would make the most sense for B to be consistent between touch and keyboard, regardless what it does or did. But I don't recall seeing anything recently that might've affected it? Maybe a regression related to full keyboard input?

**Comment by poire-z** [comment]
A is caused by dd74194e0. I guess I'll just restore it when has_pages, as I don't want to look into how koptinterface's getWordFromPosition does it :/  (I get fractional coordinates, that don't work better when I floor()/ceil() them, and which may be wrong).
```diff
--- a/frontend/apps/reader/modules/readerhighlight.lua
+++ b/frontend/apps/reader/modules/readerhighlight.lua
@@ -761,13 +761,15 @@ function ReaderHighlight:onHold(arg, ges)
         if self.ui.document.info.has_pages then
             local boxes = {}
             table.insert(boxes, self.selected_word.sbox)
             self.view.highlight.temp[self.hold_pos.page] = boxes
+            -- Unfortunately, getWordFromPosition() may not return good coordinates,
+            -- so refresh the whole page
+            UIManager:setDirty(self.dialog, "ui")
+        else
+            -- With crengine, getWordFromPosition() does return good coordinates
+            UIManager:setDirty(self.dialog, "ui", self.selected_word.sbox)
         end
-        -- Unfortunately, CREngine does not return good coordinates
-        -- UIManager:setDirty(self.dialog, "ui")
-        -- But now it does:
-        UIManager:setDirty(self.dialog, "ui", self.selected_word.sbox)
         self:_resetHoldTimer()
```

> Maybe a regression related to full keyboard input?

No, I think it never worked, there is just no handler for them in DictQuickLookup, while there are in ScrollTextWidget. Unless it was transformed into other events from somewhere else, but I doubt it :)

**Comment by sonix-github** [comment]
> Looking at A.
> 
> About B / 4:
> 
> > Press up/down HW device buttons - not possible to move to next/previous dictionary
> > Before I used koreader nightly v2020.12-30-gb78d9c1_2020-12-23 and this issues were not there present
> 
> Are you sure about that ? Because I don't see they were ever made to switch to prev/next dict (but they were made to scroll up/down inside a defintion).
> I guess they could be made to behave like that - but I want a confirmation it's not a regression :)

You are right. It's not regression. The buttons had not the functionality to move across dictionaries in mentioned koreader version. I'am sorry for incorrect information.

**Comment by poire-z** [comment]
OK, A fixed and B implemented in #7128 .
